Save high score and provide a start level as an argument

joe-herbert opened the pull request:
Saves high score in file and provides message after game ends.
Allows user to provide an integer argument which specifies the start level in order to skip the easier early levels.
Update README to reflect changes

joe-herbert committed:
Add functionality to provide a start level as a command argument

joe-herbert committed:
Add functionality to save high score in a file

joe-herbert committed:
Update README

PonasKovas commented:
Thank you for this PR! Always exciting to see something like this. Just a couple of comments before merging though.

PonasKovas commented on src/main.rs:
```diff
@@ -44,37 +47,57 @@ impl Direction {
+        //|11|12| 9|10|
+        //|  |13|14|15|
+        //-------------
+        for i in 1..(start_level + 3) {
+            initial_snake_parts.insert(((i - (i / terminal_width)) % terminal_width, i / terminal_width));
+        }
+
+        for i in 1..(start_level + 3) {
+            initial_ordered_snake_parts.push_back(((i - (i / terminal_width)) % terminal_width, i / terminal_width));
+        }
```
Do we need 2 loops here?

PonasKovas commented on src/main.rs:
```diff
@@ -44,37 +47,57 @@ impl Direction {
             snake: Snake {
-                direction: Direction::Right,
+                //if there is less than 5 cells before the snake will crash into itself then set the direction to down
+                direction: if start_level % terminal_width >= terminal_width - 5 { Direction::Down } else { Direction::Right },
```
Is this right? I think this works for the first line, but as the snake loops around the map, this doesn't really help much at higher levels. Maybe a better technique would be to always start going down, since there's almost always much space downwards?

PonasKovas commented on src/main.rs:
```diff
@@ -44,37 +47,57 @@ impl Direction {
             input: input.read_async(),
             ended: false,
-            score: 0,
+            score: u32::from(start_level),
```
`start_level as u32` would look cleaner here I think

PonasKovas commented on src/main.rs:
```diff
@@ -269,6 +298,25 @@ impl Game {
     }
+    //returns true if new high score, and the previous high score
+    fn game_finish(self: &mut Game) -> (bool, u32) {
+        //save preferably in $HOME/.snake but fallback to /etc/.snake
+        let mut home = match home::home_dir() {
+            Some(path) => path,
+            None => PathBuf::from(r"/etc/"),
+        };
```
I don't think `/etc/` is a good fallback to `$HOME`. I'd rather it just wouldn't save the highscore if `$HOME` is not defined.

PonasKovas reviewed (changes requested):
Other than that, seems good. Again, thanks for contributing.

joe-herbert committed:
Pull request changes requested
Combine two loops
Change start direction to always down
 - since we always start going down it doesn't make much sense to always
   generate the initial food to the right as this means the user has to
   turn around so the initial position is now always random
Cleaner type conversion
Don't save the high score if $HOME is not accessible
 - includes new game end message for this case

joe-herbert commented:
Commit 59f422c should fix these changes. Thanks for your comments, this PR was my first time using rust so some of the complexities of the language meant I overlooked some basic programming issues (like the two loops, that was just dumb).

I believe the start direction did work correctly on all levels, but I see what you mean about it being easier to start going down, so I've made that change. That also meant the initial food position didn't make much sense, so I made that always randomly generated.

Hope you're happy with these changes.

PonasKovas committed:
final touch up
